public/dorequest.php

```diff
@@ -277,6 +345,104 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
+            return DoRequestError('Access denied.', 405, 'access_denied');
+        }
+
+        $achievementIdsInput = request()->input('a', '');
+        $hardcore = (bool) request()->input('h', 0);
+        $validationHash = request()->input('v');
+
+        if (!$delegateTo) {
+            return DoRequestError("You must specify a target user.", 400);
+        }
+
+        if (strcasecmp($validationHash, md5($achievementIdsInput . $delegateTo . $hardcore)) !== 0) {
+            return DoRequestError('Access denied.', 403, 'access_denied');
+        }
+
+        $targetUser = User::firstWhere('User', $delegateTo);
+        if (!$targetUser) {
+            return DoRequestError("The target user couldn't be found.", 404, 'not_found');
+        }
+
+        $achievementIdsArray = explode(',', $achievementIdsInput);
+        $filteredAchievementIds = array_filter($achievementIdsArray, function ($id) {
+            return filter_var($id, FILTER_VALIDATE_INT) !== false;
+        });
+
+        // Fetch all achievements already awarded to the user.
+        $foundPlayerAchievements = PlayerAchievement::where('user_id', $targetUser->id)
+            ->with('achievement')
+            ->whereIn('achievement_id', $achievementIdsArray)
+            ->get();
+
+        $alreadyAwardedIds = [];
+
+        // Filter out achievements based on the hardcore flag and existing unlocks.
+        $filteredAchievementIds = array_filter($achievementIdsArray, function ($id) use (&$alreadyAwardedIds, $user, $foundPlayerAchievements, $hardcore) {
+            foreach ($foundPlayerAchievements as $foundPlayerAchievement) {
+                if ($foundPlayerAchievement->achievement_id == $id) {
+                    // If a record exists and it's already unlocked in hardcore mode, then
+                    // we won't attempt to re-award the achievement to the user.
+                    if ($hardcore && $foundPlayerAchievement->unlocked_hardcore_at !== null) {
+                        $alreadyAwardedIds[] = $foundPlayerAchievement->achievement_id;
+
+                        return false;
+                    }
+
+                    // If a record exists but it's a softcore unlock and hardcore mode is requested,
+                    // then we're going to add it to the list of attempted IDs to unlock.
+                    if ($hardcore && $foundPlayerAchievement->unlocked_hardcore_at === null) {
+                        return true;
+                    }
+
+                    // If a record exists (either hardcore or softcore) and softcore mode is requested,
+                    // then skip over this ID.
+                    if (!$hardcore) {
+                        $alreadyAwardedIds[] = $foundPlayerAchievement->achievement_id;
+
+                        return false;
+                    }
+
+                    if (!$foundPlayerAchievement->achievement->getCanDelegateUnlocks($user)) {
+                        return false;
+                    }
+                }
```

Comment:
recommend adding a `break` here after finding the matching record

Reply:
I went a bit further and opted to eliminate the `foreach` entirely, leaning more heavily into collections with `firstWhere()`.